### PR TITLE
WIP CLI add refactoring AEA-490

### PR DESCRIPTION
**Pull request by panasevychol**
## Proposed changes

CLI add refactoring (wip)
#997 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/cli/add.py [hunk 1]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
+
+    :return: None
+    """
     if item_type in {"connection", "skill"}:
-        _add_protocols(click_context, item_configuration.protocols)
+        _add_protocols(click_context, item_config.protocols)
 
     if item_type == "skill":
-        for contract_public_id in item_configuration.contracts:
-            if contract_public_id not in ctx.agent_config.contracts:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:196]: why has this line been removed?

File: aea/cli/add.py [hunk 2]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
+
+    :return: None.
+    """
+    fingerprint = _compute_fingerprint(
+        package_path, ignore_patterns=item_config.fingerprint_ignore_patterns
+    )
+    return item_config.fingerprint == fingerprint
+
+
+def _load_item_config(item_type: str, package_path: Path) -> ConfigLoader:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:226]: The return type should be `Configuration`
This should be a class method of `Configuration` imo. Then it is more likely to be reused.

File: aea/cli/add.py [hunk 3]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
+    ctx = cast(Context, click_context.obj)
+    for protocol_public_id in protocols:
+        if protocol_public_id not in ctx.agent_config.protocols:
+            logger.debug(
+                "Adding protocol '{}' to the agent...".format(protocol_public_id)
+            )
+            _add_item(click_context, "protocol", protocol_public_id)
+
+
+def _is_fingerprint_correct(package_path: Path, item_config) -> bool:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:211]: This should be a generic helper. So not private, and in helper/ folder

File: aea/cli/add.py [hunk 4]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
+    Add protocols to AEA by list of public IDs.
+
+    :param click_context: click context object.
+    :param protocols: a Collection of protocol public IDs to be added.
+
+    :return: None
+    """
+    ctx = cast(Context, click_context.obj)
+    for protocol_public_id in protocols:
+        if protocol_public_id not in ctx.agent_config.protocols:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:204]: Just move these two lines to 253, no need for an extra method..confuses more than it helps

File: aea/cli/add.py [hunk 5]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
+    item_config = _load_item_config(item_type, package_path)
+
+    if not _is_fingerprint_correct(package_path, item_config):
+        raise click.ClickException("Failed to add an item with incorrect fingerprint.")
+
+    _add_item_deps(click_context, item_type, item_config)
+    _register_item(ctx, item_type, item_public_id)
+
+
+def _is_item_present(ctx: Context, item_type: str, item_public_id: PublicId) -> bool:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:167]: Seems like the `Configuration` should expose a method to check if any item of item type is present in it. Then this method is no longer needed.

File: aea/cli/add.py [hunk 6]
```diff
@@ -181,58 +155,124 @@ def _add_item(click_context, item_type, item_public_id) -> None:
...
 
     if item_type == "skill":
-        for contract_public_id in item_configuration.contracts:
-            if contract_public_id not in ctx.agent_config.contracts:
-                _add_item(click_context, "contract", contract_public_id)
+        for contract_public_id in item_config.contracts:
+            _add_item(click_context, "contract", contract_public_id)
+
+
+def _register_item(ctx: Context, item_type: str, item_public_id: PublicId) -> None:
```
Comment (DavidMinarsch) [inline aea/cli/add.py:260]: `Configuration` should expose this as a method

File: aea/cli/add.py (hunk 4 above)
Comment (panasevychol) [reply aea/cli/add.py:204]: Agree. Basically no need as we already check that in _is_item_present.

File: aea/cli/add.py (hunk 5 above)
Comment (panasevychol) [reply aea/cli/add.py:167]: Requires Configuration upgrade. Currently AgentConfig (I suppose this method should belong to this class) doesn't have a property for cwd, well it seems it has `directory` property but in it's code I see that is never assigned and always None. So currently we can't check a package presence in agent folder.
I propose to leave it for now here and remove after AgentConfig upgrade so we can use a cwd from it's props (worth to create a ticket for that).
Correct me if I'm wrong.

**Pull request merged**